File: 3742-maximum-path-score-in-a-grid/solution.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int maxPathScore(vector<vector<int>>& grid, int k) {
        int m = grid.size();
        int n = grid[0].size();

        // dp[j][c] stores the maximum score at column j with cost c
        // Initializing with -1 to represent unreachable states
        vector<vector<int>> dp(n, vector<int>(k + 1, -1));

        // Starting point (0, 0) is guaranteed to be grid[0][0] == 0
        dp[0][0] = 0;

        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                int score_inc = grid[i][j];
                int cost_inc = (grid[i][j] > 0) ? 1 : 0;

                for (int c = k; c >= 0; --c) {
                    int prev_max = -1;

                    // Case 1: Coming from the top (same column, previous row)
                    // The value dp[j][c] currently holds the value from the previous row
                    if (i > 0 && c >= cost_inc) {
                        prev_max = max(prev_max, dp[j][c - cost_inc]);
                    }

                    // Case 2: Coming from the left (previous column, same row)
                    // The value dp[j-1][c] has already been updated for the current row
                    if (j > 0 && c >= cost_inc) {
                        prev_max = max(prev_max, dp[j - 1][c - cost_inc]);
                    }
                    
                    // Specific logic for (0,0) as it has no neighbors
                    if (i == 0 && j == 0) {
                        // dp[0][0] is already set to 0, which is correct
                        continue; 
                    }

                    if (prev_max != -1) {
                        dp[j][c] = prev_max + score_inc;
                    } else {
                        dp[j][c] = -1;
                    }
                }
            }
        }

        int max_score = -1;
        for (int c = 0; c <= k; ++c) {
            max_score = max(max_score, dp[n - 1][c]);
        }

        return max_score;
    }
};

File: 3742-maximum-path-score-in-a-grid/solution.py
class Solution:
    def maxPathScore(self, grid: List[List[int]], k: int) -> int:
        m, n = len(grid), len(grid[0])
        
        # dp[j][c] stores the max score reaching column j with total cost c
        # Initialized to -1 for unreachable states
        dp = [[-1] * (k + 1) for _ in range(n)]
        
        # Base case: start at (0, 0). grid[0][0] is always 0.
        dp[0][0] = 0
        
        for i in range(m):
            # We must process columns carefully to use current row vs previous row values
            for j in range(n):
                val = grid[i][j]
                cost_inc = 1 if val > 0 else 0
                score_inc = val
                
                # Special skip for the very first cell after initialization
                if i == 0 and j == 0:
                    continue
                
                # Update DP table for current cell (i, j)
                # We iterate backwards if updating in-place to avoid using the same cell twice,
                # but here we need to compare current row left neighbor and previous row top neighbor.
                for c in range(k, cost_inc - 1, -1):
                    res = -1
                    # From top (previous row, same column)
                    if i > 0:
                        res = max(res, dp[j][c - cost_inc])
                    # From left (same row, previous column)
                    if j > 0:
                        res = max(res, dp[j-1][c - cost_inc])
                    
                    if res != -1:
                        dp[j][c] = res + score_inc
                    else:
                        dp[j][c] = -1
                
                # If cost_inc was 1, cost 0 is no longer reachable at this cell
                if cost_inc > 0:
                    dp[j][0] = -1
                    
        ans = max(dp[n-1])
        return ans if ans >= 0 else -1